README.md
## `survivalMPLdc`: Survival Analysis under Dependent Right Censoring using Maximum Penalised Likelihood

## survivalMPLdc
The `survivalMPLdc` package provides the estimating and plotting function for the Cox hazard models under dependent right censoring using maximum penalised likelihood method [Xu et al (2018)](https://onlinelibrary.wiley.com/doi/abs/10.1002/sim.7651).

The `survivalMPLdc` currently only supports proportional hazard models. The future work can extend the proposed method to other semi-parametric hazard model, i.e. the accelerate failure or additive hazard models. Other type of censoring, i.e. interval censoring and even time-dependent covariates can be considered. 

## Installation

### Stable release on CRAN

The `survivalMPLdc` package can be installed from `CRAN`.

```s
install.packages("survivalMPLdc")
library("survivalMPLdc")
```

### Development version on Github

You can use the devtools package to install the development version of `survivalMPLdc` from GitHub:
```s
# install.packages("devtools")
devtools::install_github("Kenny-Jing-Xu/survivalMPLdc")
library(survivalMPLdc)
```

## Usage
A reference manual is available at [kenny-jing-xu.github.io/survivalMPLdc](https://kenny-jing-xu.github.io/survivalMPLdc/).

## Citation
Xu, J., Ma, J. and Fung, T. (2020). survivalMPLdc: Survival Analysis under Dependent Right Censoring using Maximum Penalised Likelihood. R package version 0.1.0.


NEWS.md

## survivalMPLdc 0.1.0

* First major release of the package.


tests/testthat/test_survivalMPLdc.R
context("Testing the regression cofficients estimation")


library(survivalMPLdc)
library(copula)
library(survival)
library(splines2)
library(matrixcalc)
library(testthat)

test_that("Testing the MPL estimated regression coefficients versus the true values", {

##-- Copula types
copula3 <- 'frank'

##-- Marginal distribution for T, C, and A
a <- 2
lambda <- 2
cons7 <- 0.2
cons9 <- 10
tau <- 0.8
betas <- c(-0.5, 0.1)
phis <- c(0.3, 0.2)
distr.ev <- 'weibull'
distr.ce <- 'exponential'

##-- Real marginal baseline hazard for T
ht0b <- a * (seq(0, 5.4, 0.01) ^ (a - 1)) / (lambda ^ a)

##-- Sample size 200
n <- 5000
set.seed(0)
##-- One sample Monte Carlo dataset
cova <- cbind(rbinom(n, 1, 0.5), runif(n, min=-10, max=10))
set.seed(0)
surv <- surv_data_dc(n, a, cova, lambda, betas, phis, cons7, cons9, tau, copula3,
                      distr.ev, distr.ce)
n <- nrow(cova)
p <- ncol(cova)

##-- event and dependent censoring proportions
colSums(surv)[c(2,3)]/n
X <- surv[,1] # Observed time
del<-surv[,2] # failure status
eta<-surv[,3] # dependent censoring status



##-- Selecting bin sample or number of knots using AIC vs binCounts plot
binCounts <- c(500, 1000, 1500, 2000, 2500)
bn <- length(binCounts)
aics<-rep(0, bn)
for (j in 1:bn)
{ control=coxph_mpl_dc.control(ordSp = 4,
                               binCount = binCounts[j], tie = 'No',
                               tau = 0.8, copula = copula3,
                               pent = 'penalty_mspl', smpart = 0, penc = 'penalty_mspl', smparc = 0,
                               maxit2 = 100, maxit = 5000,
                               mid = 1, asy = 0, ac = 1, cv = 0,
                               ac.theta = 1e-5, ac.gamma = 1e-5, ac.Utheta = -1e-2, ac.Ugamma = -1e-2,
                               min.theta = 1e-7, min.gamma = 1e-7, min.ht = 1e-7, min.hc = 1e-7,
                               min.St = 1e-7, min.Sc = 1e-7, min.C = 1e-7, min.dC = 1e-7,
                               eps = 1e-5, tol.thga = 1e-5, tol.bph = 1e-5, tol.smpar = 1e-2,
                               cat.smpar = 'No')
aics[j]<-coxph_mpl_dc(surv, cova, control)$mpl_aic
#print(j)
}
binCount <- binCounts[ which.min( aics ) ]
plot(binCounts, aics)

coxMPLests <- coxph_mpl_dc(surv, cova, ordSp=4, binCount=binCount, tie='No', tau=0.8, copula=copula3,
                            pent='penalty_mspl', smpart='REML', penc='penalty_mspl', smparc='REML',
                            maxit2=100, maxit=10000, mid=1, asy=1, ac=1, cv=1,
                            ac.theta=1e-5, ac.gamma=1e-5, ac.Utheta=-1e-2, ac.Ugamma=-1e-2,
                            min.theta=1e-7, min.gamma=1e-7, min.ht=1e-7, min.hc=1e-7,
                            min.St=1e-7, min.Sc=1e-7,
                            min.C=1e-7, min.dC=1e-7, eps=1e-5,
                            tol.thga=1e-5, tol.bph=1e-5, tol.smpar=1e-2,
                            cat.smpar='No'
)
mpl_beta_phi_zp <- rbind( coef(object = coxMPLests, parameter = "beta",),
                          coef(object = coxMPLests, parameter = "phi",)
)


##-- Plot the true and estimated baseline hazards (95% confidence interval) for T
plot(x = coxMPLests, parameter = "theta", funtype="hazard",
     xout = seq(0, 5.4, 0.01), se = TRUE,
     cols=c("blue", "red"), ltys=c(4, 2), type="l", lwd=1, cex=1, cex.axis=1, cex.lab=1,
     xlab="Time", ylab="Hazard",
     xlim=c(0, 5.4), ylim=c(0, 4)
)
par(new=TRUE)
plot(seq(0, 5.4, 0.01), ht0b,
     type="l", col="green",
     lty=1, lwd=1, cex.axis=1, cex.lab=1, xlim=c(0, 5.4), ylim=c(0, 4),
     xlab='Time', ylab='Hazard')
title("MPL Hazard", cex.main=1)
legend( 'topleft',legend = c( "MPL", "95% Confidence Interval", "True"),
        col = c("blue", "red", "green"),
        lty = c(4, 2, 1),
        cex = 1)

expect_equal(
  round(aics,2),
  c(12772.98, 12766.93, 12761.98, 12759.81, 12755.97)
)

expect_equal(
  round( mpl_beta_phi_zp[,1], 8 ),
  c(-0.49690106, 0.09671272, 0.33270738, 0.19981402)
  )

expect_equal(
  round( mpl_beta_phi_zp[,2], 8 ),
  c(0.033302473, 0.003093199, 0.037263795, 0.004401592)
)

})




R/PRIME.R
#' PRIME data set
#'
#' This data set is from a longitudinal study called "Prospective Research in Memory Clinics" (PRIME),
#' see Brodaty et al (2011), with a period of 3-year. The data set includes 583 dementia patients. The
#' outcome is time to institutionalised. The predicators are age, sex, educational level, living status,
#' dementia type, baseline cognitive ability (MMSE), baseline functional ability (SMAF),
#' baseline neuropsychiatric symptoms (total NPI), baseline dementia severity (CDR),
#' baseline caregiver burden (ZBI), medication types, change in cognitive ability at 3 months,
#' change in functional ability at 3 months, and change in neuropsychiatric symptoms at 3 months.
#' Note that this data set is complete and was analyzed by Brodaty et al (2014).
#'
#' @name PRIME
#' @format A data frame with 583 observations on 18 variables.
#' \describe{
#' \item{Time}{observed time in term of months}
#' \item{Event}{event or institutionalisation, 1=Yes and 0=No}
#' \item{Depcen}{dependent right censoring or withdrawal, 1=Yes and 0=No}
#' \item{Age}{age at baseline, 1=80 years or above and 0=80 year below}
#' \item{Gender}{gender, 1=Female and 0=Male}
#' \item{HighEdu}{education level at baseline, 1=high school above and 0=high school
#'                or below}
#' \item{Alzheimer}{Alzheimer disease, 1=Yes and 0=No}
#' \item{CDR_base}{dementia severity at baseline}
#' \item{MMSE_base}{cognitive ability at baseline}
#' \item{SMAF_base}{functional ability at baseline}
#' \item{ZBI_base}{caregiver burden at baseline}
#' \item{NPI_base}{neuropsychiatric symptoms at baseline}
#' \item{Benzon}{benzodiazepines taking, 1=Yes and 0=No}
#' \item{Antiphsy}{anti-psychotics taking, 1=Yes and 0=No}
#' \item{LivingAlone}{living alone, 1=Yes and 0=No}
#' \item{MMSE_change_3m}{cognitive ability change at 3-month from baseline}
#' \item{SMAF_change_3m}{functional ability change at 3-month from baseline}
#' \item{NPI_change_3m}{neuropsychiatric symptoms change at 3-month from baseline}
#' }
#'
#' @docType data
#' @keywords datasets
#' @usage
#' data
#' @references
#' Brodaty H, Woodward M, Boundy K, Ames D, Balshaw R. (2011). \emph{"Patients in Australian memory clinics:
#' baseline characteristics and predictors of decline at six months"}. Int Psychogeriatr 23, 1086-1096.
#'
#' Brodaty H, Connors M, Xu J, Woodward M, Ames D. (2014). \emph{"Predictors of institutionalization
#' in dementia: a three year longitudinal study"}. Journal of Alzheimers Disease 40, 221-226.
NULL



R/surv_data_dc.R

#' @title Generate a sample of time to event dataset with dependent right censoring under an Archimedean copula
#'
#' @description Generate a sample of time to event dataset with, dependent right censoring based on one of the Archimedean copulas the given Kendall's tau, sample size n and covariates matrix Z.
#'
#' @usage surv_data_dc(n, a, Z, lambda, betas, phis, cons7, cons9, tau, copula, distr.ev, distr.ce)
#'
#' @param n the sample size, or the number of the subjects in a sample.
#' @param a the shape parameter of baseline hazard for the event time \eqn{T}.
#' @param Z the covariate matrix with dimension of \eqn{n} by \eqn{p}, where \eqn{p} is the number of covariates.
#' @param lambda the scale parameter of baseline hazard for event time \eqn{T}.
#' @param betas the regression coefficient vector of proportional hazard model for the event time \eqn{T} with dimenion of \eqn{p} by \eqn{1}.
#' @param phis the regression coefficient vector of proportional hazard model for dependent censoring time \eqn{C} with dimenion of \eqn{p} by \eqn{1}.
#' @param cons7 the parameter of baseline hazard for the dependent censoring time \eqn{C} if assuming an exponential distribution.
#' @param cons9 the upper limit parameter of uniform distribution for the independent censoring time \eqn{A},
#' i.e. \eqn{A}~U(0, cons9).
#' @param tau the Kendall's correlation coefficient between \eqn{T} and \eqn{C}.
#' @param copula the Archemedean copula that captures the dependence between \eqn{T} and \eqn{C}, a characteristc value,
#' i.e. 'independent', 'clayton', 'gumbel' or 'frank'.
#' @param distr.ev the distribution of the event time, a characteristc value, i.e. 'weibull' or 'log logit'.
#' @param distr.ce the distribution of the dependent censoring time, a characteristc value, i.e. 'exponential' or 'weibull'.
#'
#' @details surv_data_dc allows to generate a survival dataset under dependent right censoring, at sample size \code{n}, based on one of the Archimedean \code{copula},
#' Kendall's \code{tau}, and covariates matrix \code{Z} with dimension of \eqn{n} by \eqn{p}. For example, at \code{p=2}, we have \code{Z=cbind(Z1, Z2)},
#' where \code{Z1} is treatment generated by distribution of bernoulli(0.5), i.e. 1 represents treatment group and 0 represents control group; \code{Z2} is the age
#' generated by distribution of U(-10, 10).
#'
#' The generated dataset includes three varaibles, which are \eqn{X_{i}}, \eqn{\delta_{i}} and \eqn{\eta_i}, i.e. \eqn{X_i=\text{min}(T_i, C_i, A_i)},
#' \eqn{\delta_i=I(X_i=T_i)} and \eqn{\eta_i=I(X_i=C_i)}, for \eqn{i=1,\ldots,n}. 'T' represents the event time, whose hazard function is
#' \deqn{h_{T}(x)=h_{0T}(x)\text{exp}(\boldsymbol Z^{\top}\beta)}, where the baseline hazard can take weibull form, i.e. \eqn{h_{0T}(x) = ax^{a-1} / \lambda^a}, or
#' log logistic form, i.e.
#' \deqn{ h_{0T}(x) = \frac{ \frac{ 1 }{ a\text{exp}( \lambda ) } \left ( \frac{ x }{ \text{exp}( \lambda ) } \right)^{1/a -1 } }{ 1 + \left( \frac{ x }{ \text{exp}( \lambda ) } \right)^{1/a}  } }.
#' 'C' represents the dependent censoring time, whose hazard function is \eqn{ h_{C}(x) = h_{0C}(x)\text{exp}(\boldsymbol Z^{\top}\phi) }, where the baseline hazard can take exponential form, i.e.
#'  \eqn{h_{0C}(x)=cons7}, or weibull form, i.e. \eqn{h_{0C}(x) = ax^{a-1} / \lambda^a}.'A' represents the administrative or independent censoring time, where A~U(0, cons9).
#'
#' @return A sample of time to event dataset under dependent right censoring, which includes observed time \eqn{X}, event indicator \eqn{\delta} and dependent censoring indicator \eqn{\eta}.
#'
#' @author Jing Xu, Jun Ma, Thomas Fung
#'
#' @references Xu J, Ma J, Connors MH, Brodaty H. (2018). \emph{"Proportional hazard model estimation under dependent censoring using copulas and penalized likelihood"}.
#' Statistics in Medicine 37, 2238–2251.
#'
#' @seealso \code{\link{coxph_mpl_dc}}
#'
#' @import copula
#'
#' @importFrom stats pnorm qexp qweibull runif
#'
#' @examples
#'  library(copula)
#'
#'  ##-- Copula types
#'  copula1 <- 'clayton'
#'  copula2 <- 'gumbel'
#'  copula3 <- 'frank'
#'  copula4 <- 'independent'
#'
#'  ##-- Marginal distribution for T, C, and A
#'  a <- 2
#'  lambda <- 2
#'  cons7 <- 0.2
#'  cons9 <- 10
#'  tau <- 0.8
#'  betas <- c(-0.5, 0.1)
#'  phis <- c(0.3, 0.2)
#'  distr.ev <- 'weibull'
#'  distr.ce <- 'exponential'
#'
#'  ##-- Sample size
#'  n <- 200
#'
#'  ##-- One sample Monte Carlo dataset
#'  cova <- cbind(rbinom(n, 1, 0.5), runif(n, min=-10, max=10))
#'  surv <- surv_data_dc(n, a, cova, lambda, betas, phis, cons7, cons9,
#'                      tau, copula3, distr.ev, distr.ce)
#'  n <- nrow(cova)
#'  p <- ncol(cova)
#'  ##-- event and dependent censoring proportions
#'  colSums(surv)[c(2,3)]/n
#'  X <- surv[,1] # Observed time
#'  del<-surv[,2] # failure status
#'  eta<-surv[,3] # dependent censoring status
#'
#'
#' @export

surv_data_dc<-function(n, a, Z, lambda, betas, phis, cons7, cons9, tau, copula, distr.ev, distr.ce)
{
  {
    if(copula=='independent')
    {
      cop=indepCopula(2)
    }
    else if(copula=='clayton')
    {
      alpha<-iTau(claytonCopula(100), tau)
      cop <- claytonCopula(alpha, dim = 2)
    }
    else if(copula=='gumbel')
    {
      alpha<-iTau(gumbelCopula(100), tau)
      cop <- gumbelCopula(alpha, dim = 2)
    }
    else if(copula=='frank')
    {
      alpha<-iTau(frankCopula(100), tau)
      cop <- frankCopula(alpha, dim = 2)
    }
    else (stop("Error! Define a proper copula function"))
  }
  x<-rep(0, n)
  del<-rep(0, n)
  eta<-rep(0, n)
  expbetT=exp(Z%*%betas)
  lamT<-lambda*(exp(-Z%*%betas))^(1/a)
  lamCe<-cons7*exp(Z%*%phis)
  lamCw<-lambda*(exp(-Z%*%phis))^(1/a)
  for (i in (1:n))
  {
    cuv<-rCopula(1, cop)      #generate the values of marginal suvival functions
    if( distr.ev == 'weibull' )
    {
      ti<-qweibull((1-cuv[1,1]), a, lamT[i])   #weibull distribution
    }
    else if ( distr.ev == 'log logit' )
    {
      ti=exp( lambda )*( cuv[1,1]^( -1/expbetT[i] ) - 1 )^( a )     #log logit distribution
    }
    else ( stop("Error! Define a proper distribution type for the event time") )
    if( distr.ce=='exponential' )
    {
      ci<-qexp((1-cuv[1,2]), lamCe[i]) #exponential distribution
    }
    else if ( distr.ce=='weibull' ) {
      ci<-qweibull((1-cuv[1,2]), a, lamCw[i])   #weibull distribution
    }
    else ( stop("Error! Define a proper distribution type for the dependent censoring time") )
    ai<-runif(1, min=0, max=cons9)
    if(ti<ci & ti<ai){
      del[i]<-1
      x[i]<-ti
    }
    else if (ci<ti & ci <ai)
    {
      eta[i]<-1
      x[i]<-ci
    }
    else{
      del[i]<-0
      eta[i]<-0
      x[i]<-ai
    }
  }
  dat<-cbind(x, del, eta)
  return(dat)
}






R/survivalMPLdc-package.R
#' Survival analysis under dependent right censoring using maximum Penalised
#' likelihood method
#'
#' @name survivalMPLdc-package
#' @aliases survivalMPLdc
#' @docType package
#' @title Survival analysis under dependent right censoring using maximum Penalised
#' likelihood method
#' @keywords package
#' @references
#' Ma, J. (2010). \emph{"Positively constrained multiplicative iterative algorithm
#' for maximum penalised likelihood tomographic reconstruction"}. IEEE Transactions On Signal Processing
#' 57, 181-192.
#'
#' Brodaty H, Woodward M, Boundy K, Ames D, Balshaw R. (2011). \emph{"Patients in Australian memory clinics:
#' baseline characteristics and predictors of decline at six months"}. Int Psychogeriatr 23, 1086-1096.
#'
#' Ma, J. and Heritier, S. and Lo, S. (2014). \emph{"On the Maximum Penalised Likelihood Approach
#' for Proportional Hazard Models with Right Censored Survival Data"}. Computational Statistics and
#' Data Analysis 74, 142-156.
#'
#' Brodaty H, Connors M, Xu J, Woodward M, Ames D. (2014). \emph{"Predictors of institutionalization
#' in dementia: a three year longitudinal study"}. Journal of Alzheimers Disease 40, 221-226.
#'
#' Xu J, Ma J, Connors MH, Brodaty H. (2018). \emph{"Proportional hazard model estimation under
#' dependent censoring using copulas and penalized likelihood"}. Statistics in Medicine 37, 2238–2251.
NULL


R/plot.coxph_mpl_dc.R
#' @title Plot a baseline hazard estimates from coxph_mpl_dc Object
#'
#' @description Plot the baseline hazard with the confidence interval estimates
#'
#' @method plot coxph_mpl_dc
#'
#' @param x an object inheriting from class \code{\link{coxph_mpl_dc}}
#' @param parameter the set of parameters of interest.  Indicate \code{parameters="theta"} for the baseline hazard
#' estimated by \eqn{theta} and \code{parameters="gamma"} for the baseline hazard estimated by \eqn{gamma}
#' @param funtype the type of function for ploting, i.e. \code{funtype="hazard"} for baseline hazard,
#' \code{funtype="cumhazard"} for baseline cumulative hazard and \code{funtype="survival"} for
#' baseline survival function
#' @param xout the time values for the baseline hazard plot
#' @param se se=TRUE gives both the MPL baseline estimates and 95\% confidence interval plots while se=FALSE
#' gives only the MPL baseline estimate plot.
#' @param ltys a line type vector with two components, the first component is the line type of the baseline hazard
#' while the second component is the line type of the 95\% confidence interval
#' @param cols a colour vector with two components, the first component is the colour of the baseline hazard
#' while the second component is the colour the 95\% confidence interval
#' @param ... other arguments
#'
#'
#' @details When the input is of class \code{coxph_mpl_dc} and \code{parameters=="theta"}, the baseline estimates
#' base on \eqn{theta} and xout (with the corresponding 95\% confidence interval if se=TRUE ) are ploted.
#' When the input is of class \code{coxph_mpl_dc} and \code{parameters=="gamma"}, the baseline hazard estimates
#' based on \eqn{gamma} and xout (with the corresponding 95\% confidence interval if se=TRUE ) are ploted.
#'
#' @return the baseline hazard plot
#'
#' @author Jing Xu, Jun Ma, Thomas Fung
#'
#' @references Brodaty H, Connors M, Xu J, Woodward M, Ames D. (2014). \emph{"Predictors of institutionalization
#' in dementia: a three year longitudinal study"}. Journal of Alzheimers Disease 40, 221-226.
#'
#' Xu J, Ma J, Connors MH, Brodaty H. (2018). \emph{"Proportional hazard model estimation under dependent censoring using copulas and penalized likelihood"}.
#' Statistics in Medicine 37, 2238–2251.
#'
#'
#' @seealso \code{\link{coef.coxph_mpl_dc}}, \code{\link{coxph_mpl_dc.control}}, \code{\link{coxph_mpl_dc}}
#'
#' @import copula
#'
#' @importFrom survival coxph
#' @importFrom splines2 iSpline mSpline
#' @importFrom graphics par
#' @importFrom stats approx
#'
#' @examples
#' \donttest{
#'  library(copula)
#'  library(survival)
#'  library(splines2)
#'
#'  ##-- Copula types
#'  copula1 <- 'clayton'
#'  copula2 <- 'gumbel'
#'  copula3 <- 'frank'
#'  copula4 <- 'independent'
#'
#' ##-- A real example
#' ##-- One dataset from Prospective Research in Memory Clinics (PRIME) study
#' ##-- Refer to article Brodaty et al (2014),
#' ##   the predictors of institutionalization of dementia patients over 3-year study period
#'
#' data(PRIME)
#'
#' surv=as.matrix(PRIME[,1:3]) #time, event and dependent censoring indicators
#' cova=as.matrix(PRIME[, -c(1:3)]) #covariates
#' colMeans(surv[,2:3])  #the proportions of event and dependent censoring
#'
#' n=dim(PRIME)[1];print(n)
#' p=dim(PRIME)[2]-3;print(p)
#' names(PRIME)
#'
#' ##-- Selecting number of knots under independent censoring assumption using AIC vs binCounts plot
#' min.binCount=20
#' max.binCount=200
#' incr.binCount=20
#' binCounts <- seq(min.binCount, max.binCount, incr.binCount)
#' bn <- length(binCounts)
#' aics<-rep(0, bn)
#'
#' for (j in 1:bn)
#' { control=coxph_mpl_dc.control(ordSp = 4,
#'                                binCount = binCounts[j], tie = 'Yes',
#'                                tau = 0, copula = copula4,
#'                                pent = 'penalty_mspl', smpart = 0,
#'                                penc = 'penalty_mspl', smparc = 0,
#'                                maxit2 = 100, maxit = 5000,
#'                                mid = 1, asy = 0, ac = 1, cv = 0,
#'                                ac.theta = 1e-5, ac.gamma = 1e-5,
#'                                ac.Utheta = -1e-2, ac.Ugamma = -1e-2,
#'                                min.theta = 1e-7, min.gamma = 1e-7,
#'                                min.ht = 1e-7, min.hc = 1e-7,
#'                                min.St = 1e-7, min.Sc = 1e-7, min.C = 1e-7, min.dC = 1e-7,
#'                                eps = 1e-5, tol.thga = 1e-5, tol.bph = 1e-5, tol.smpar = 1e-2 )
#' aics[j]<-coxph_mpl_dc(surv, cova, control)$mpl_aic
#' print(j) }
#' binCount = min.binCount + ( which.min( aics ) - 1 )*incr.binCount
#' plot(binCounts, aics)

#' ##--MPL estimate Cox proportional hazard model for institutionalization under medium positive
#' ##--dependent censoring
#' control <- coxph_mpl_dc.control(ordSp = 4,
#'                                 binCount = binCount, tie = 'Yes',
#'                                 tau = 0.5, copula = copula3,
#'                                 pent = 'penalty_mspl', smpart = 'REML',
#'                                 penc = 'penalty_mspl', smparc = 'REML',
#'                                 maxit2 = 100, maxit = 5000,
#'                                 mid = 1, asy = 1, ac = 1, cv = 1,
#'                                 ac.theta = 1e-5, ac.gamma = 1e-5,
#'                                 ac.Utheta = -1e-2, ac.Ugamma = -1e-2,
#'                                 min.theta = 1e-7, min.gamma = 1e-7, min.ht = 1e-7, min.hc = 1e-7,
#'                                 min.St = 1e-7, min.Sc = 1e-7, min.C = 1e-7, min.dC = 1e-7,
#'                                 eps = 1e-5, tol.thga = 1e-5, tol.bph = 1e-5, tol.smpar = 1e-2 )
#'
#' coxMPLests_tau <- coxph_mpl_dc(surv=surv, cova=cova, control=control, )
#'
#' plot(x = coxMPLests_tau, parameter = "theta", funtype="hazard",
#'      xout = seq(0, 36, 0.01), se = TRUE,
#'      cols=c("blue", "red"), ltys=c(1, 2), type="l", lwd=1, cex=1, cex.axis=1, cex.lab=1,
#'      xlab="Time (Month)", ylab="Hazard",
#'      xlim=c(0, 36), ylim=c(0, 0.05)
#'      )
#'      title("MPL Hazard", cex.main=1)
#'      legend( 'topleft',legend = c( expression(tau==0.5), "95% Confidence Interval"),
#'      col = c("blue", "red"),
#'      lty = c(1, 2),
#'      cex = 1)
#'
#'plot(x = coxMPLests_tau, parameter = "theta", funtype="cumhazard",
#'     xout = seq(0, 36, 0.01), se = TRUE,
#'     cols=c("blue", "red"), ltys=c(1, 2),
#'     type="l", lwd=1, cex=1, cex.axis=1, cex.lab=1,
#'     xlab="Time (Month)", ylab="Hazard",
#'     xlim=c(0, 36), ylim=c(0, 1.2)
#')
#'title("MPL Cumulative Hazard", cex.main=1)
#'legend( 'topleft',
#'        legend = c( expression(tau==0.5), "95% Confidence Interval"),
#'        col = c("blue", "red"),
#'        lty = c(1, 2),
#'        cex = 1
#')
#'
#'plot(x = coxMPLests_tau, parameter = "theta", funtype="survival",
#'     xout = seq(0, 36, 0.01), se = TRUE,
#'     cols=c("blue", "red"), ltys=c(1, 2),
#'     type="l", lwd=1, cex=1, cex.axis=1, cex.lab=1,
#'     xlab="Time (Month)", ylab="Hazard",
#'     xlim=c(0, 36), ylim=c(0, 1)
#')
#'title("MPL Survival", cex.main=1)
#'legend( 'bottomleft',
#'        legend = c( expression(tau==0.5), "95% Confidence Interval"),
#'        col = c("blue", "red"),
#'        lty = c(1, 2),
#'        cex = 1
#')
#'
#'}
#'
#' @export

plot.coxph_mpl_dc<-function(x, parameter="theta", funtype="hazard", xout, se=TRUE, ltys, cols, ...)
{
  surv = x$surv
  Xi=surv[,1]
  max_Xi=max(Xi)

  if(missing(parameter)){parameter="theta"}
  if(missing(funtype)){funtype="hazard"}
  if(missing(xout)){ xout = seq(0, max_Xi, 0.01) }
  if(missing(se)){se=FALSE}
  if(missing(cols)){cols=c("black", "black")}
  if(missing(ltys)){ltys=c(1, 2)}

  if(parameter=="theta")
  {
    if(funtype=="hazard"){

      mpl_h0 <- x$mpl_h0t
      mpl_h0_sd = x$mpl_ht0_sd
      mpl_h0i <- approx( Xi, mpl_h0, xout = xout,
                         method="constant", rule = 2, ties = mean)$y
      mpl_h0i_sd <- approx( Xi, mpl_h0_sd, xout = xout,
                            method="constant", rule = 2, ties = mean)$y
      yout = mpl_h0i
      youtu = mpl_h0i + 1.96*mpl_h0i_sd
      youtl = mpl_h0i - 1.96*mpl_h0i_sd
      youtl[youtl<0]=0

    }else if(funtype=="cumhazard"){

      mpl_H0 <- x$mpl_H0t
      mpl_H0_sd = x$mpl_Ht0_sd
      mpl_H0i <- approx( Xi, mpl_H0, xout = xout,
                         method="constant", rule = 2, ties = mean)$y
      mpl_H0i_sd <- approx( Xi, mpl_H0_sd, xout = xout,
                            method="constant", rule = 2, ties = mean)$y
      yout = mpl_H0i
      youtu = mpl_H0i + 1.96*mpl_H0i_sd
      youtl = mpl_H0i - 1.96*mpl_H0i_sd
      youtl[youtl<0]=0

      }else if(funtype=="survival"){

        mpl_S0 <- x$mpl_S0t
        mpl_S0_sd = x$mpl_St0_sd
        mpl_S0i <- approx( Xi, mpl_S0, xout = xout,
                           method="constant", rule = 2, ties = mean)$y
        mpl_S0i_sd <- approx( Xi, mpl_S0_sd, xout = xout,
                              method="constant", rule = 2, ties = mean)$y
        yout = mpl_S0i
        youtu = mpl_S0i + 1.96*mpl_S0i_sd
        youtl = mpl_S0i - 1.96*mpl_S0i_sd
        youtl[youtl<0]=0
        youtu[youtu>1]=1

        }else (stop("Error! Define a proper function type"))

  }else if(parameter=="gamma"){

    if(funtype=="hazard"){

      mpl_h0 <- x$mpl_h0c
      mpl_h0_sd = x$mpl_hc0_sd
      mpl_h0i <- approx( Xi, mpl_h0, xout = xout,
                         method="constant", rule = 2, ties = mean)$y
      mpl_h0i_sd <- approx( Xi, mpl_h0_sd, xout = xout,
                            method="constant", rule = 2, ties = mean)$y
      yout = mpl_h0i
      youtu = mpl_h0i + 1.96*mpl_h0i_sd
      youtl = mpl_h0i - 1.96*mpl_h0i_sd
      youtl[youtl<0]=0

    }else if(funtype=="cumhazard"){

      mpl_H0 <- x$mpl_H0c
      mpl_H0_sd = x$mpl_Hc0_sd
      mpl_H0i <- approx( Xi, mpl_H0, xout = xout,
                         method="constant", rule = 2, ties = mean)$y
      mpl_H0i_sd <- approx( Xi, mpl_H0_sd, xout = xout,
                            method="constant", rule = 2, ties = mean)$y
      yout = mpl_H0i
      youtu = mpl_H0i + 1.96*mpl_H0i_sd
      youtl = mpl_H0i - 1.96*mpl_H0i_sd
      youtl[youtl<0]=0

    }else if(funtype=="survival"){

      mpl_S0 <- x$mpl_S0c
      mpl_S0_sd = x$mpl_Sc0_sd
      mpl_S0i <- approx( Xi, mpl_S0, xout = xout,
                         method="constant", rule = 2, ties = mean)$y
      mpl_S0i_sd <- approx( Xi, mpl_S0_sd, xout = xout,
                            method="constant", rule = 2, ties = mean)$y
      yout = mpl_S0i
      youtu = mpl_S0i + 1.96*mpl_S0i_sd
      youtl = mpl_S0i - 1.96*mpl_S0i_sd
      youtl[youtl<0]=0
      youtu[youtu>1]=1

    }else (stop("Error! Define a proper function type"))


  }else (stop("Error! Define a proper parameter name"))

  if(se==TRUE){

    plot( x = xout, y = yout, col=cols[1], lty=ltys[1],... )
    par(new=TRUE)
    plot( x = xout, y = youtu, col=cols[2], lty=ltys[2],... )
    par(new=TRUE)
    plot( x = xout, y = youtl, col=cols[2], lty=ltys[2],... )

  }else if(se==FALSE){

    plot( x = xout, y = yout, col=cols[1], lty=ltys[1],... )

  }

}
